Fine-grained Authorization for ResourceClaim Status Updates

aojea opened the pull request:
### What this PR does / why we need it:

This PR implements KEP-4817, introducing fine-grained authorization checks for Dynamic Resource Allocation (DRA) ResourceClaim status updates.

Currently, granting a DRA plugin or controller permission to update resourceclaims/status allows it to modify any field in the status, including allocations and device statuses belonging to other drivers or nodes. To improve security and enforce least privilege, this PR locks down these updates behind a new feature gate: DRAResourceClaimGranularStatusAuthorization (Beta, enabled by default in 1.36).

### How it works:

We inject the API Authorizer into the ResourceClaim REST storage strategy and evaluate granular permissions during ValidateUpdate:

1. Allocations (status.allocation, status.reservedFor): The caller now needs permission to update the synthetic resourceclaims/binding subresource.

2. Device Status (status.devices): The caller needs permission to update the synthetic resourceclaims/driver subresource.

To prevent a compromised node from altering another node's devices, we evaluate this using specialized synthetic verb prefixes: associated-node:<verb> (if the caller is a ServiceAccount on the node where the claim is allocated) or arbitrary-node:<verb> (for control-plane controllers).

### Example for DRA Plugin Developers:

Third-party plugin developers will need to update their ClusterRole or Role manifests to grant explicit permissions for their specific driver name on the new resourceclaims/driver subresource.

Here is an example of what a Kubelet plugin's RBAC will look like moving forward (assuming the plugin runs on the node and updates status via PATCH or UPDATE):

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: my-dra-plugin-role
rules:
# 1. Standard permission to update the status subresource
- apiGroups: ["resource.k8s.io"]
  resources: ["resourceclaims/status"]
  verbs: ["get", "patch", "update"]

# 2. NEW: Granular permission to update devices specific to this driver
- apiGroups: ["resource.k8s.io"]
  resources: ["resourceclaims/driver"]
  # Use 'associated-node:' prefix for node-local plugins, or 
  # 'arbitrary-node:' for central control-plane controllers
  verbs: ["associated-node:patch", "associated-node:update"]
  resourceNames: ["my-dra-driver.example.com"] 
```

### Which issue(s) this PR fixes:


```release-note

ACTION REQUIRED: DRA (Dynamic Resource Allocation) drivers and controllers now require granular RBAC permissions to update ResourceClaim statuses when the `DRAResourceClaimGranularStatusAuthorization` feature gate is enabled (Beta in 1.36). Schedulers and controllers must be granted `update`/`patch` on `resourceclaims/binding`. DRA drivers must be granted `associated-node:update` or `arbitrary-node:update` (or patch equivalents) on `resourceclaims/driver`, restricted by their specific `resourceNames`.
```

enj commented:
@aojea please move this logic into the strategy of the resource like we did for pod certs.

The authorization checks are not optional and cannot be disabled.  The cluster admin can always choose to grant this access to `system:authenticated` if they do not wish to have granular enforcement, but they cannot prevent the check from running.  This keeps the behavior in all clusters identical.

https://github.com/kubernetes/kubernetes/blob/a8a43a5e250368aefb7cf11c4932193257961b59/pkg/registry/certificates/podcertificaterequest/strategy.go#L148-L164

cc @deads2k @liggitt

aojea commented on plugin/pkg/admission/resourceclaim/admission.go:
```diff
@@ -0,0 +1,286 @@
+			APIGroup:        driverGroupResource.Group,
+			Resource:        driverGroupResource.Resource,
+			ResourceRequest: false, // This implies name="*"
+		}
```
Thanks appreciate the guidance

aojea committed:
Add granular authorization for DRA ResourceClaim status updates

This commit introduces the DRAResourceClaimGranularStatusAuthorization
feature gate (Beta in 1.36) to enforce fine-grained authorization checks
on ResourceClaim status updates.

Previously, 'update' permission on 'resourceclaims/status' allowed modifying
the entire status. To enforce the principle of least privilege for DRA
drivers and the scheduler, this change introduces synthetic subresources and
verb prefixes:

- 'resourceclaims/binding': Required to update 'status.allocation' and
  'status.reservedFor'.
- 'resourceclaims/driver': Required to update 'status.devices'. Evaluated
  on a per-driver basis using 'associated-node:<verb>' (for node-local
  ServiceAccounts) or 'arbitrary-node:<verb>' (for cluster-wide controllers).

aojea commented:
heh , it seems to work

> { failed [FAILED] ResourceClaim.resource.k8s.io "external-claim" is invalid: status.allocation.devices: Forbidden: user "system:serviceaccount:dra-8583:dra-kubelet-plugin-dra-8583.k8s.io-service-account" is not authorized to update device status for driver "dra-8583.k8s.io"
In [It] at: k8s.io/kubernetes/test/e2e/dra/dra.go:2664 @ 10/30/25 21:43:30.829

thockin reviewed (commented):
It's still early in the declarative validation process, but what I REALLY want to see, as a project, is a shift away from "let's write some bespoke apiserver code for this" to "let's declare the semantics alongside the schema, and let the apiserver implement it" -- CRDs and builtins need to get closer to each others' capabilities.

aojea commented on pkg/registry/resource/utils.go:
```diff
@@ -102,3 +108,148 @@ func adminRequested(deviceRequestResults []resource.DeviceRequestAllocationResul
+// AuthorizedForDeviceStatus checks if the request status is authorized update the device status
+func AuthorizedForDeviceStatus(ctx context.Context, authz authorizer.Authorizer, newAllocatedDeviceStatus, oldAllocatedDeviceStatus resource.ResourceClaimStatus, namespaceName string) field.ErrorList {
+	var allErrs field.ErrorList
+	fieldPath := field.NewPath("status", "allocation", "devices")
```
I can pass the fieldPath as argument

aojea commented on test/e2e/dra/dra.go:
```diff
@@ -2690,6 +2692,152 @@ var _ = framework.SIGDescribe("node")(framework.WithLabel("DRA"), func() {
 			gomega.Expect(getResourceClaim).ToNot(gomega.BeNil())
 			gomega.Expect(getResourceClaim.Status.Devices).To(gomega.Equal(updatedResourceClaim.Status.Devices))
 		})
+
```
@pohly bringing the discussion https://github.com/kubernetes/kubernetes/pull/134947#issuecomment-3479122775 here, the test `must be possible for the driver to update the ResourceClaim.Status.Devices once allocated` is specific of a node local driver that attaches an interface that has IP (think on secondary nics for pods) . The test fails because of https://github.com/kubernetes/kubernetes/pull/134947/files#r2484831605 

From your comment it is unclear to me if that the test is wrong, as it does not emulate a driver in the node? 

The test fails because

aojea replied:
done

seans3 commented:
/remove-sig api-machinery

aojea commented:
this was demanded by sig auth for the graduation, so I think the ball is on them to review @ritazh @enj do you mind allocating some time to unblock this? we postponed the promotion one release because of your feedback

enj reviewed (changes requested):
First pass.

aramase commented on pkg/registry/resource/utils.go:
```diff
@@ -102,3 +110,151 @@ func adminRequested(deviceRequestResults []resource.DeviceRequestAllocationResul
 }
+
+// AuthorizedForDeviceStatus checks if the request status is authorized update the device status
+func AuthorizedForDeviceStatus(ctx context.Context, fieldPath *field.Path, authz authorizer.Authorizer, newAllocatedDeviceStatus, oldAllocatedDeviceStatus resource.ResourceClaimStatus, namespaceName string) field.ErrorList {
```
`namespaceName` is never used?

aramase commented on pkg/registry/resource/utils.go:
```diff
@@ -102,3 +110,151 @@ func adminRequested(deviceRequestResults []resource.DeviceRequestAllocationResul
+	}
+
+	// Check authorization for the specific driver name
+	for _, driverName := range driversToAuthz.UnsortedList() {
```
```suggestion
	for _, driverName := range sets.List(driversToAuthz) {
```

`UnsortedList()` causes non-deterministic errors. When multiple drivers fail authz, error order is random.

aojea replied:
TIL

aojea commented on pkg/registry/resource/utils.go:
```diff
@@ -102,3 +110,151 @@ func adminRequested(deviceRequestResults []resource.DeviceRequestAllocationResul
+
+		// If the request is from a node, do not fall back to wildcard permission,
+		// nodes must have explicit permission for each driver they report status for.
+		if isNode {
```
I mean, it is whatever the helper of the first commit defines

aojea replied:
fixed , removed the helper and restricteied to service accounts only

aojea commented:
> [It] [sig-apps] StatefulSet Basic StatefulSet functionality [StatefulSetBasic] should perform rolling updates with maxUnavailable [FeatureGate:MaxUnavailableStatefulSet] [Beta] 

/test pull-kubernetes-e2e-kind

> {Failed  === RUN   TestCustomResourceValidatorsWithSchemaConversion
    testserver.go:628: Resolved testserver package path to: "/home/prow/go/src/k8s.io/kubernetes/cmd/kube-apiserver/app/testing"
    testserver.go:426: runtime-config=map[]
    testserver.go:427: Starting kube-apiserver on port 46627...
    testserver.go:462: Waiting for /healthz to be ok...
    crd_validation_expressions_test.go:722: expect error to contain 
ule compiler initialization error: failed to convert to declType for CEL validation rules" but get: <nil>

/test pull-kubernetes-integration

does not look related

enj commented on pkg/registry/resource/utils.go:
```diff
@@ -102,3 +112,182 @@ func adminRequested(deviceRequestResults []resource.DeviceRequestAllocationResul
+	}
+
+	associatedNodeName, isAssociatedWithNode := nodeNameFromNodeBoundToken(user)
+	// If the request is from a node, check if the claim is allocated to that node,
```
```suggestion
	// If the request is associated with a node, check if the claim is allocated to that node,
```

enj reviewed (changes requested):
Another pass.

enj reviewed (commented):
Minor stuff.
